File: python_leaoai/repositories/Factories/Propriedade/TipoCategoriaFactory.py
from sqlalchemy.orm import Session
from typing import List, Optional
from data.models.propriedade import TipoCategoria


class TipoCategoriaFactory:
    def __init__(self, db: Session):
        self.db = db

    # -------------------------
    # CRUD
    # -------------------------

    def get_obj(self, id: int) -> Optional[TipoCategoria]:
        return self.db.query(TipoCategoria).filter(TipoCategoria.id == id).first()

    def save_obj(self, obj: TipoCategoria) -> TipoCategoria:
        self.db.add(obj)
        self.db.commit()
        self.db.refresh(obj)
        return obj

    def update_obj(self, obj: TipoCategoria) -> TipoCategoria:
        self.db.merge(obj)
        self.db.commit()
        return obj

    def get_all(self) -> List[TipoCategoria]:
        return self.db.query(TipoCategoria).all()

    def get_all_ativo(self) -> List[TipoCategoria]:
        return self.db.query(TipoCategoria).filter(TipoCategoria.ativo.is_(True)).all()

    def get_all_excluido(self) -> List[TipoCategoria]:
        return self.db.query(TipoCategoria).filter(TipoCategoria.excluido.is_(True)).all()

    def remove_obj(self, obj: TipoCategoria) -> bool:
        try:
            self.db.delete(obj)
            self.db.commit()
            return True
        except Exception:
            self.db.rollback()
            return False

    def delete_obj(self, id: int) -> bool:
        try:
            obj = self.get_obj(id)
            if obj is None:
                return False

            obj.ativo = False
            obj.excluido = True
            self.db.merge(obj)
            self.db.commit()
            return True
        except Exception:
            self.db.rollback()
            return False

    # -------------------------
    # Extra
    # -------------------------

    def get_obj_by_url(self, url: str) -> Optional[TipoCategoria]:
        return (
            self.db.query(TipoCategoria)
            .filter(TipoCategoria.ativo.is_(True), TipoCategoria.url == url)
            .first()
        )

File: Framework.Repositories/Factories/Propriedade/TipoCategoriaFactory.cs
using Data;
using Data.Models.Propriedade;
using System.Collections.Generic;
using System.Linq;

namespace Framework.Factories.Propriedade
{
    public class TipoCategoriaFactory
    {
        private readonly ApplicationDbContext _context;

        public TipoCategoriaFactory(ApplicationDbContext context)
        {
            _context = context;
        }

        #region Crud

        public TipoCategoria GetObj(int id)
        {
            return _context.TipoCategoria.Where(b => b.Id == id).FirstOrDefault();
        }
        public TipoCategoria SaveObj(TipoCategoria obj)
        {
            _context.Add(obj);
            _context.SaveChanges();

            return obj;
        }

        public TipoCategoria UpdateObj(TipoCategoria obj)
        {
            _context.Update(obj);
            _context.SaveChanges();
            return obj;
        }

        public List<TipoCategoria> GetAll()
        {
            return _context.TipoCategoria.ToList();
        }

        public List<TipoCategoria> GetAllAtivo()
        {
            return _context.TipoCategoria.Where(b=> b.Ativo == true).ToList();
        }

        public List<TipoCategoria> GetAllExcluido()
        {
            return _context.TipoCategoria.Where(b => b.Excluido == true).ToList();
        }

        public Boolean RemoveObj(TipoCategoria obj)
        {
            try
            {
                _context.TipoCategoria.Remove(obj);
                _context.SaveChanges();

                return true;
            }
            catch (Exception)
            {
                return false;
            }
        } 

        public Boolean DeleteObj(int Id)
        {
            try
            {
                var obj =_context.TipoCategoria.Where(b => b.Id == Id).FirstOrDefault();

                obj.Ativo = false;
                obj.Excluido = true;

                _context.TipoCategoria.Update(obj);
                _context.SaveChanges();
                
                return true;
            }
            catch (Exception)
            {
                return false;
            }
        }
        #endregion

        public TipoCategoria GetObjByUrl(String url)
        {
            return _context.TipoCategoria.Where(b =>b.Ativo == true && b.Url == url).FirstOrDefault();
        }
      
    }
}
